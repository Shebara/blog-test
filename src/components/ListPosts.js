import React, { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import { Link } from 'react-router-dom';
import { setData } from '../store/actions';
import { saveData } from '../store/localStorage';

function mapDispatchToProps(dispatch) {
    return {
        setData: data => dispatch(setData(data))
    };
}

class ListPosts extends Component {
    constructor(props) {
        super(props);

        this._isMounted = false;
        this.cancelToken = false;
    }

    componentDidMount() {
        this._isMounted = true;
    }

    async appendCommentCount(data) {
        const vm = this;
        const url = 'http://localhost:5000/comments/' + data.id + '/?nocache=' + new Date().getTime();
        const response = await axios.get(url, {
            cancelToken: new axios.CancelToken(function executor(cancel) {
                vm.cancelToken = cancel;
            })
        }).catch(function (thrown) {
            if (axios.isCancel(thrown)) {
              console.log('Request `' + url + '` interrupted.');
            }
        });

        const myComments = this.props.comments.filter(item => item.postId === data.id);

        data.commentCount = response && response.data && response.data.length > 0
        ? response.data.length + myComments.length : 0 + myComments.length;

        return data;
    }

    async getResponse() {
        const vm = this;
        const url = 'http://localhost:5000/post/?nocache=' + new Date().getTime();
        const response = await axios.get(url, {
            cancelToken: new axios.CancelToken(function executor(cancel) {
                vm.cancelToken = cancel;
            })
        }).catch(function (thrown) {
            if (axios.isCancel(thrown)) {
              console.log('Request `' + url + '` interrupted.');
            }
        });

        if (! this._isMounted || ! response) {
            return;
        }

        let data = response.data;

        if (data && ! data.message) {
            data = await Promise.all(data.map(item => this.appendCommentCount(item)))

            this.props.setData({ data });
            saveData('data', data);
        }
    }

    renderItems() {
        const data = this.props.data;

        return (
            <ul>
                {data.map((item) => {
                    const link = '/post/' + item.id;
                    return (
                        <li key={item.id}>
                            <Link to={link}>
                                <div className="box" title={item.title}>
                                    <h4>{item.title}</h4>
                                    <p>{item.description}</p>
                                    <div className="created-at">{item.createdAt}</div>
                                    <div className="comments">{item.commentCount} comments</div>
                                </div>
                            </Link>
                        </li>
                    );
                })}
            </ul>
        )
    }

    render() {
        if (this.props.data.length === 0) {
            this.getResponse();

            return (
                <div>
                    <h3>Loading posts...</h3>
                </div>
            );
        }

        return (
            <div className="ListPosts">
                {this.renderItems()}
            </div>
        );
    }

    componentWillUnmount() {
        if (this.cancelToken) {
            this.cancelToken();
        }

        this._isMounted = false;
    }
}


export default connect(null, mapDispatchToProps)(ListPosts);
