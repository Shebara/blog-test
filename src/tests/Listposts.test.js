import React from 'react';
import { shallow } from 'enzyme';
import ListPosts from '../components/ListPosts';

describe('ListPosts', () => {
    it('should render without errors', () => {
        const component = shallow(<ListPosts />);

        expect(component.getElements()).toMatchSnapshot();
    });
});
