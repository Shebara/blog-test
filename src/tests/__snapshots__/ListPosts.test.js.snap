// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListPosts should render without errors 1`] = `
Array [
  <Context.Provider
    value={
      Object {
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "subscription": Subscription {
          "handleChangeWrapper": [Function],
          "listeners": Object {
            "notify": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": null,
        },
      }
    }
  >
    <UNDEFINED />
  </Context.Provider>,
]
`;
